backend/app/anti_recommenders/arkg/arkg_anti_recommender.py

```diff
@@ -0,0 +1,162 @@
+import typing
+from collections.abc import Iterable
+from pathlib import Path
+
+import pyoxigraph as ox
+
+from app.anti_recommenders import AntiRecommender
+from app.constants import WIKIPEDIA_BASE_URL
+from app.models import AntiRecommendation, AntiRecommendationGraph
+from app.models.types import RdfMimeType, RecordKey, StoreQuery
+from app.namespaces import SCHEMA
+
+
+class ArkgAntiRecommender(AntiRecommender):
+    """
+    A concrete implementation of AntiRecommender.
+
+    ArkgAntiRecommender uses information stored in an Anti-Recommendation Knowledge Graph to generate AntiRecommendations.
+    """
+
+    def __init__(
+        self,
+        *,
+        arkg_base_iri: ox.NamedNode,
+        arkg_file_path: Path,
+        arkg_mime_type: RdfMimeType,
+        record_keys: tuple[RecordKey, ...],
+    ) -> None:
+        self.__anti_recommendation_path_graph: list[AntiRecommendationGraph] = []
+        self.__arkg_base_iri = arkg_base_iri
+        self.__arkg_file_path = arkg_file_path
+        self.__arkg_mime_type = arkg_mime_type
+        self.__record_keys: tuple[RecordKey, ...] = record_keys
+        self.__store: ox.Store = ox.Store()
+        self.__load_store()
+
+    def __create_query(self, record_key: RecordKey) -> StoreQuery:
+        """
+        Return a SPARQL query that fetches anti-recommendations of `record_key` from an ARKG Store.
+        """
+
+        return f"SELECT ?title WHERE {{ <{record_key}> <{SCHEMA.ITEM_REVIEWED.value}> ?resource {{?resource <{SCHEMA.TITLE.value}> ?title}} }}"
+
+    def __build_anti_recommendation_path_graph(
```

Comment:
Rather than building a redundant data structure, use a database query and keep minimal state in the code.

Reply:
All the state you really need outside the graph is:
* a set of node identifiers the user has already seen, to prevent cycles
* the identifier of the last node the user saw, so you know where to start the graph query

You could store both of them in the graph, but graphs aren't really meant for storing user data in a transactional way. I'd investigate Supabase, which can do arbitrary tables (it's built on Postgres) but also has authentication (e.g., OAuth) and user management.

Reply:
Essentially you need user management + some minimal per-user data. All of the BaaS will support that. I'd encapsulate it in an API in order to stay vendor-neutral. Supabase (self-hosted), Firebase, AWS Cognito, Auth0.

Reply:
I redesigned ArkgAntiRecommender and added a Supabase database. I am yet to add authentication, but the ArkgAntiRecommender can now keep track of per-user data (via an API).